igvc_navigation/src/mapper/map_utils.cpp

```diff
@@ -0,0 +1,231 @@
+        filtered_pc.points.emplace_back(i);
+      }
+    }
+    else
+    {
+      filtered_pc.points.emplace_back(i);
+    }
+  }
+}
+
+std::optional<GroundPlane> filterGroundPlane(const PointCloud& raw_pc, PointCloud& ground, PointCloud& nonground,
+                                             GroundFilterOptions options)
+{
+  ground.header = raw_pc.header;
+  nonground.header = raw_pc.header;
+
+  std::optional<GroundPlane> ground_plane = ransacFilter(raw_pc, ground, nonground, options.ransac_options);
+  if (ground_plane)
+  {
+    return ground_plane;
+  }
+  else
+  {
+    fallbackFilter(raw_pc, ground, nonground, options.fallback_options);
+    return std::nullopt;
+  }
+}
+
+void projectTo2D(PointCloud& projected_pc)
+{
+  for (auto& point : projected_pc.points)
+  {
+    point.z = 0;
+  }
+}
+
+std::optional<GroundPlane> ransacFilter(const PointCloud& raw_pc, PointCloud& ground, PointCloud& nonground,
+                                        RANSACOptions options)
+{
+  // Plane detection for ground removal
+  pcl::ModelCoefficientsPtr coefficients(new pcl::ModelCoefficients);
+  pcl::PointIndices::Ptr inliers(new pcl::PointIndices);
+
+  // create the segmentation object
+  pcl::SACSegmentation<pcl::PointXYZ> seg;
+  seg.setOptimizeCoefficients(true);
+
+  seg.setModelType(pcl::SACMODEL_PERPENDICULAR_PLANE);
+  seg.setMethodType(pcl::SAC_RANSAC);
+  seg.setMaxIterations(options.iterations);
+  seg.setDistanceThreshold(options.distance_threshold);
+  seg.setAxis(Eigen::Vector3f(0, 0, 1));
+  seg.setEpsAngle(options.eps_angle);
+
+  PointCloud::Ptr cloud_filtered = raw_pc.makeShared();
+
+  // Create filtering object
+  pcl::ExtractIndices<pcl::PointXYZ> extract;
+
+  seg.setInputCloud(cloud_filtered);
+  seg.segment(*inliers, *coefficients);
+
+  if (inliers->indices.empty())
+  {
+    return std::nullopt;
+  }
+  else
+  {
+    ROS_DEBUG("Ground plane found: %zu/%zu inliers. Coeff: %f %f %f %f", inliers->indices.size(),
+              cloud_filtered->size(), coefficients->values.at(0), coefficients->values.at(1),
+              coefficients->values.at(2), coefficients->values.at(3));
+    extract.setInputCloud(cloud_filtered);
+    extract.setIndices(inliers);
+    extract.setNegative(false);
+    extract.filter(ground);
+
+    // remove ground points from full pointcloud
+    if (inliers->indices.size() != cloud_filtered->size())
+    {
+      extract.setNegative(true);
+      PointCloud out;
+      extract.filter(out);
+      nonground += out;
+      *cloud_filtered = out;
+    }
+
+    return GroundPlane{ coefficients->values.at(0), coefficients->values.at(1), coefficients->values.at(2),
+                        coefficients->values.at(3) };
+  }
+}
+
+void fallbackFilter(const PointCloud& raw_pc, PointCloud& ground, PointCloud& nonground, FallbackOptions options)
+{
+  pcl::PassThrough<pcl::PointXYZ> fallback;
+  fallback.setFilterFieldName("z");
+  fallback.setFilterLimits(-options.plane_distance, options.plane_distance);
```

Comment:
Wait doesn't this filter the ground plane relative to the height of the LiDAR? not base_footprint?

I would expect something like this: 

```
// z -> z translation component from LiDAR to base_link (negative in this case)
fallback.setFilterLimits(z - options.plane_distance, z + options.plane_distance);
```

Reply:
this is done relative to odom, since the cloud is transformed before this is run on it.